v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/PipelineController.java

```diff
@@ -220,88 +102,94 @@ static PipelineResult executeShardedMigration(
+
+        // If a namespace is configured for a shard uses that, otherwise uses the namespace
+        // configured in the options if there is one.
+        String namespace = Optional.ofNullable(shard.getNamespace()).orElse(options.getNamespace());
+
+        ShardedDbConfigContainer dbConfigContainer =
+            new ShardedDbConfigContainer(
+                shard, sqlDialect, namespace, shardId, entry.getKey(), options);
+        setupLogicalDbMigration(
+            options,
+            pipeline,
+            spannerConfig,
+            tableSelector,
+            levelToSpannerTableList,
+            dbConfigContainer);
       }
     }
     return pipeline.run();
   }
 
-  // Calculate the total number of logical shards in the list of physical shards.
-  private static long findNumLogicalshards(List<Shard> shards) {
-    return shards.stream().mapToLong(shard -> shard.getDbNameToLogicalShardIdMap().size()).sum();
+  private static void setupLogicalDbMigration(
+      SourceDbToSpannerOptions options,
+      Pipeline pipeline,
+      SpannerConfig spannerConfig,
+      TableSelector tableSelector,
+      Map<Integer, List<String>> levelToSpannerTableList,
+      DbConfigContainer configContainer) {
+    Map<Integer, PCollection<Void>> levelVsOutputMap = new HashMap<>();
+    for (int currentLevel = 0; currentLevel < levelToSpannerTableList.size(); currentLevel++) {
+      List<String> spannerTables = levelToSpannerTableList.get(currentLevel);
+      LOG.info("processing level: {} spanner tables: {}", currentLevel, spannerTables);
+      List<String> sourceTables =
+          spannerTables.stream()
+              .map(t -> tableSelector.getSchemaMapper().getSourceTableName("", t))
+              .collect(Collectors.toList());
+      LOG.info("level: {} source tables: {}", currentLevel, spannerTables);
+      PCollection<Void> previousLevelPCollection = levelVsOutputMap.get(currentLevel - 1);
+      if (currentLevel > 0 && previousLevelPCollection == null) {
+        LOG.warn(
+            "proceeding without waiting for parent. current level: {}  tables: {}",
+            currentLevel,
+            spannerTables);
+      }
+      OnSignal<@UnknownKeyFor @Nullable @Initialized Object> waitOnSignal =
+          previousLevelPCollection != null ? Wait.on(previousLevelPCollection) : null;
+      JdbcIoWrapper jdbcIoWrapper =
+          JdbcIoWrapper.of(configContainer.getJDBCIOWrapperConfig(sourceTables, waitOnSignal));
+      if (jdbcIoWrapper.getTableReaders().isEmpty()) {
+        LOG.info("not creating reader as tables are not found at source: {}", sourceTables);
+        // If tables of 1 level are ignored in middle, then the subsequent level will not wait to
+        // begin processing.
+        continue;
+      }
+      ReaderImpl reader = ReaderImpl.of(jdbcIoWrapper);
+      String suffix = generateSuffix(configContainer.getShardId(), currentLevel + "");
+      String shardIdColumn = ""; // TODO- Check if this is needed
```

Comment:
We seem to be passing 2 shard id columns now in MigrateTableTransform: shardId and shardIdColumn. THe usage of the latter seems like former should suffice? @VardhanThigle could you verify?

Reply:
I don't think the shardId column is needed here. Hence the TODO - if it is confirmed we can remove it.

Reply:
The `shardIdColumn` is needed for DLQ retrials.

Reply:
As discussed, will take this up as a follow on PR.